libs/etcd_broker/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+            }
+            SubscriptionKind::Timeline(ZTenantTimelineId {
+                tenant_id,
+                timeline_id,
+            }) => format!(
+                "{}/{tenant_id}/{timeline_id}/safekeeper",
+                self.broker_prefix
+            ),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum SubscriptionKind {
+    /// Get every timeline update.
+    All,
+    /// Get certain tenant timelines' updates.
+    Tenant(ZTenantId),
+    /// Get certain timeline updates.
+    Timeline(ZTenantTimelineId),
+}
+
+/// Creates a background task to poll etcd for timeline updates from safekeepers.
+/// Stops and returns `Err` on any error during etcd communication.
+/// Watches the key changes until either the watcher is cancelled via etcd or the subscription cancellation handle,
+/// exiting normally in such cases.
+pub async fn subscribe_to_timeline_updates(
+    client: &mut Client,
+    subscription: SkTimelineSubscriptionKind,
+) -> anyhow::Result<SkTimelineSubscription> {
+    info!("Subscribing to timeline updates, subscription kind: {subscription:?}");
+
+    let (watcher, mut stream) = client
+        .watch(
+            subscription.watch_key(),
+            Some(WatchOptions::new().with_prefix()),
+        )
+        .await
+        .with_context(|| format!("Failed to init the watch for subscription {subscription:?}"))?;
+
+    let (timeline_updates_sender, safekeeper_timeline_updates) = mpsc::unbounded_channel();
+
+    let subscription_kind = subscription.kind;
+    let regex = subscription.watch_regex();
+    let watcher_handle = tokio::spawn(async move {
+        while let Some(resp) = stream.message().await.with_context(|| {
+            format!(
+                "Failed to get messages from the subscription stream, kind: {subscription_kind:?}"
+            )
+        })? {
+            if resp.canceled() {
+                info!("Watch for timeline updates subscription was canceled, exiting");
+                break;
+            }
+
+            let mut timeline_updates: HashMap<ZTenantTimelineId, HashMap<ZNodeId, SkTimelineInfo>> =
+                HashMap::new();
+
+            let events = resp.events();
+            debug!("Processing {} events", events.len());
+
+            for event in events {
+                if EventType::Put == event.event_type() {
+                    if let Some(kv) = event.kv() {
+                        match parse_etcd_key_value(subscription_kind, &regex, kv) {
+                            Ok(Some((zttid, timeline))) => {
+                                match timeline_updates
+                                    .entry(zttid)
+                                    .or_default()
+                                    .entry(timeline.safekeeper_id)
+                                {
+                                    hash_map::Entry::Occupied(mut o) => {
+                                        if o.get().flush_lsn < timeline.info.flush_lsn {
```

Comment:
I'm not sure about this, because in theory `flush_lsn` is not always a correct metric for identifying latest update. More correct metric would be `(last_log_term, commit_lsn, flush_lsn)`, but I'd not use it because it's not very obvious. Can we just send all updates with `Some(zttid, timeline)` to a channel one by one, or there will be performance and other kind of issues?

Reply:
We can send those and I'm sure this code is more like a precaution rather then a real situation with etcd: AFAIK, per subscription poll on the key, it returns the "aggregated" data with the last versions per key value (node id in our case).

Feels odd to panic/error in such case though, so maybe just merge it over the triplet?

Reply:
https://etcd.io/docs/v3.5/learning/api_guarantees/#isolation-level-and-consistency-of-replicas

> etcd does not ensure linearizability for watch operations. Users are expected to verify the revision of watch responses to ensure correct ordering.

Looks like we have to check versions ourselves, so I'll add some code and comments there.